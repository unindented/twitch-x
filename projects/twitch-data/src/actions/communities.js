import {getAction} from '../api'

export function loadTopCommunities ({limit}) {
  const action = {
    type: 'LOAD_TOP_COMMUNITIES'
  }
  return getAction(action, '/communities/top', {query: {limit}})
}
