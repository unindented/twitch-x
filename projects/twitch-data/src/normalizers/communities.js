import {normalize, schema} from 'normalizr'
import {valueOrNull} from './utils'

const community = new schema.Entity('communities', {}, {
  idAttribute: '_id',
  processStrategy: (community) => {
    return {
      id: valueOrNull(community._id),
      name: valueOrNull(community.display_name),
      image: valueOrNull(community.avatar_image_url),
      viewers: valueOrNull(community.viewers),
      channels: valueOrNull(community.channels)
    }
  }
})

const root = [community]

export function normalizeTopCommunities (payload) {
  return normalize(payload.communities, root)
}
