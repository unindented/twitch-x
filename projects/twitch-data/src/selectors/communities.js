import {createSelector} from 'reselect'

export const getTopCommunities = createSelector(
  (state) => state.data.communities.top,
  (state) => state.data.communities.byId,
  (topCommunityIds, communitiesById) => topCommunityIds.map((id) => communitiesById[id])
)

export const getCommunity = createSelector(
  (state, props) => state.data.communities.byId[props.id],
  (community) => community
)
