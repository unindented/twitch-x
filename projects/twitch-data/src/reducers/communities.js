import update from 'immutability-helper'
import {normalizeTopCommunities} from '../normalizers/communities'

const initialState = {
  byId: {},
  top: []
}

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case 'LOAD_TOP_COMMUNITIES_SUCCESS': {
      const {entities, result} = normalizeTopCommunities(action.payload.response)
      return update(state, {$merge: {byId: entities.communities, top: result}})
    }
  }

  return state
}
