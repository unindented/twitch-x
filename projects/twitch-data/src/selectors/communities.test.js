import {getTopCommunities, getCommunity} from './communities'

describe('selectors/communities', () => {
  describe('.getTopCommunities', () => {
    it('returns the top communities when available', () => {
      const state = {
        data: {
          communities: {
            byId: {
              1: {name: 'foo'},
              2: {name: 'bar'},
              3: {name: 'baz'}
            },
            top: [1, 2]
          }
        }
      }
      expect(getTopCommunities(state)).toMatchSnapshot()
    })
  })

  describe('.getCommunity', () => {
    it('returns the community with the specified ID', () => {
      const state = {
        data: {
          communities: {
            byId: {
              1: {name: 'foo'}
            }
          }
        }
      }
      expect(getCommunity(state, {id: 1})).toMatchSnapshot()
    })
  })
})
