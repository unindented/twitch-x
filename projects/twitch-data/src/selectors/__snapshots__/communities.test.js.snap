// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/communities .getCommunity returns the community with the specified ID 1`] = `
Object {
  "name": "foo",
}
`;

exports[`selectors/communities .getTopCommunities returns the top communities when available 1`] = `
Array [
  Object {
    "name": "foo",
  },
  Object {
    "name": "bar",
  },
]
`;
